tests/sample_size/test_metrics.py

```diff
@@ -76,6 +77,74 @@ def test_boolean_metric_get_probability_too_small(self):
         )
 
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_two_variants_true_null(self, mock_nct, mock_uniform):
+        true_alt = False
+        size = 200
+        variants = 2
+        pval = 1
+        mock_uniform.return_value = pval
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_called_once_with(0, 1, size=variants - 1)
+        mock_nct.assert_not_called()
+        self.assertEqual(p, pval)
+
+    @patch("sample_size.metrics.BooleanMetric.variance")
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_two_variants_true_alt(self, mock_nct, mock_uniform, mock_variance):
+        true_alt = True
+        size = 200
+        variants = 2
+        z = 0
+        pval = 1
+        nc = np.sqrt(size / 2) * self.DEFAULT_MDE / self.DEFAULT_MOCK_VARIANCE
+        mock_variance.__get__ = MagicMock(return_value=self.DEFAULT_MOCK_VARIANCE)
+        mock_nct.return_value = z
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_not_called()
+        mock_nct.assert_called_once_with(nc=nc, df=2 * (size - 1), size=variants - 1)
+        self.assertEqual(p, pval)
+
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_multiple_variants_true_null(self, mock_nct, mock_uniform):
+        true_alt = False
+        size = 200
+        variants = 5
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        pval = np.ones(variants - 1)
+        mock_uniform.return_value = pval
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_called_once_with(0, 1, size=variants - 1)
+        mock_nct.assert_not_called()
+        self.assertTrue((p == pval).all())
+
+    @patch("sample_size.metrics.BooleanMetric.variance")
+    @patch("scipy.stats.uniform.rvs")
+    @patch("scipy.stats.nct.rvs")
+    def test_boolean_metric_generate_p_value_multiple_variants_true_alt(self, mock_nct, mock_uniform, mock_variance):
+        true_alt = True
+        size = 200
+        variants = 5
+        z = np.zeros(variants - 1)
+        pval = np.ones(variants - 1)
+        nc = np.sqrt(size / 2) * self.DEFAULT_MDE / self.DEFAULT_MOCK_VARIANCE
+        mock_variance.__get__ = MagicMock(return_value=self.DEFAULT_MOCK_VARIANCE)
+        mock_nct.return_value = z
+
+        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
+        p = boolean.generate_p_value(true_alt, size, variants)
+        mock_uniform.assert_not_called()
+        mock_nct.assert_called_once_with(nc=nc, df=2 * (size - 1), size=variants - 1)
+        self.assertTrue((p == pval).all())
```

Comment:
Might be a good place to parameterize (same goes for other classes as well)

```suggestion
    @parameterized.expand(product([True, False], [2, 5], [200]))
    @patch("scipy.stats.uniform.rvs")
    @patch("scipy.stats.nct.rvs")
    @patch("scipy.stats.norm.sf")
    def test_generate_p_value(self, mock_sf, mock_nct, mock_uniform, is_null, variants, size):
        p_value_generator = mock_uniform if is_null else mock_sf
        pval = MagicMock()
        p_value_generator.return_value = pval
        
        boolean = BooleanMetric(self.DEFAULT_PROBABILITY, self.DEFAULT_MDE)
        boolean.variance = self.DEFAULT_MOCK_VARIANCE
        p = boolean.generate_p_value(true_alt, size, variants)
        
        if is_null:
            mock_uniform.assert_called_once_with(0, 1, size=variants - 1)
            mock_nct.assert_not_called()
        else:
             mock_uniform.assert_not_called()
             mock_nct.assert_called_once_with(nc=nc, df=2 * (size - 1), size=variants - 1)
        self.assertTrue((p == pval).all())
```